File: RxTools-library/src/main/java/com/vondear/rxtools/view/RxBarCode.java
package com.vondear.rxtools.view;

import android.graphics.Bitmap;
import android.support.annotation.NonNull;
import android.widget.ImageView;

import com.google.zxing.BarcodeFormat;
import com.google.zxing.MultiFormatWriter;
import com.google.zxing.WriterException;
import com.google.zxing.common.BitMatrix;

/**
 * Created by vondear on 2017/2/17.
 */

public class RxBarCode {

    /**
     * 获取建造者
     *
     * @param text 样式字符串文本
     * @return {@link RxBarCode.Builder}
     */
    public static RxBarCode.Builder builder(@NonNull CharSequence text) {
        return new RxBarCode.Builder(text);
    }

    public static class Builder {

        private int backgroundColor = 0xffffffff;

        private int codeColor = 0xff000000;

        private int codeWidth = 1000;

        private int codeHeight = 300;

        private CharSequence content;

        public Builder backColor(int backgroundColor) {
            this.backgroundColor = backgroundColor;
            return this;
        }

        public Builder codeColor(int codeColor) {
            this.codeColor = codeColor;
            return this;
        }

        public Builder codeWidth(int codeWidth) {
            this.codeWidth = codeWidth;
            return this;
        }

        public Builder codeHeight(int codeHeight) {
            this.codeHeight = codeHeight;
            return this;
        }

        public Builder(@NonNull CharSequence text) {
            this.content = text;
        }

        public Bitmap into(ImageView imageView) {
            Bitmap bitmap = RxBarCode.createBarCode(content, codeWidth, codeHeight, backgroundColor, codeColor);
            if (imageView != null) {
                imageView.setImageBitmap(bitmap);
            }
            return bitmap;
        }
    }

    //----------------------------------------------------------------------------------------------以下为生成二维码算法

    public static Bitmap createBarCode(CharSequence content, int BAR_WIDTH, int BAR_HEIGHT, int backgroundColor, int codeColor) {
        /**
         * 条形码的编码类型
         */
        BarcodeFormat barcodeFormat = BarcodeFormat.CODE_128;
        final int backColor = backgroundColor;
        final int barCodeColor = codeColor;

        MultiFormatWriter writer = new MultiFormatWriter();
        BitMatrix result = null;
        try {
            result = writer.encode(content + "", barcodeFormat, BAR_WIDTH, BAR_HEIGHT, null);
        } catch (WriterException e) {
            e.printStackTrace();
        }

        int width = result.getWidth();
        int height = result.getHeight();
        int[] pixels = new int[width * height];
        // All are 0, or black, by default
        for (int y = 0; y < height; y++) {
            int offset = y * width;
            for (int x = 0; x < width; x++) {
                pixels[offset + x] = result.get(x, y) ? barCodeColor : backColor;
            }
        }
        Bitmap bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
        bitmap.setPixels(pixels, 0, width, 0, 0, width, height);
        return bitmap;
    }

    //----------------------------------------------------------------------------------------------生成条形码开始

    /**
     * 生成条形码
     *
     * @param contents      需要生成的内容
     * @param desiredWidth  生成条形码的宽带
     * @param desiredHeight 生成条形码的高度
     * @return
     */
    public static Bitmap createBarCode(String contents, int desiredWidth, int desiredHeight) {
        return createBarCode(contents, desiredWidth, desiredHeight, 0xFF000000, 0xFFFFFFFF);
    }

    /**
     * 生成条形码
     * desiredWidth  生成条形码的宽带
     * desiredHeight 生成条形码的高度
     *
     * @param contents 需要生成的内容
     * @return 条形码的Bitmap
     */
    public static Bitmap createBarCode(String contents) {
        return createBarCode(contents, 1000, 300);
    }

    public static void createBarCode(String content, int codeWidth, int codeHeight, ImageView iv_code) {
        iv_code.setImageBitmap(createBarCode(content, codeWidth, codeHeight));
    }

    public static void createBarCode(String content, ImageView iv_code) {
        iv_code.setImageBitmap(createBarCode(content));
    }

    //==============================================================================================生成条形码结束
}

File: RxFeature/src/main/java/com/tamsiree/rxfeature/tool/RxBarCode.kt
package com.tamsiree.rxfeature.tool

import android.graphics.Bitmap
import android.widget.ImageView
import com.google.zxing.BarcodeFormat
import com.google.zxing.MultiFormatWriter
import com.google.zxing.WriterException
import com.google.zxing.common.BitMatrix

/**
 * @author tamsiree
 * @date 2017/2/17
 */
object RxBarCode {
    /**
     * 获取建造者
     *
     * @param text 样式字符串文本
     * @return [RxBarCode.Builder]
     */
    fun builder(text: CharSequence): Builder {
        return Builder(text)
    }

    //----------------------------------------------------------------------------------------------以下为生成二维码算法
    fun createBarCode(content: CharSequence, BAR_WIDTH: Int, BAR_HEIGHT: Int, backgroundColor: Int, codeColor: Int): Bitmap {
        /**
         * 条形码的编码类型
         */
        val barcodeFormat = BarcodeFormat.CODE_128
        val writer = MultiFormatWriter()
        var result: BitMatrix? = null
        try {
            result = writer.encode(content.toString() + "", barcodeFormat, BAR_WIDTH, BAR_HEIGHT, null)
        } catch (e: WriterException) {
            e.printStackTrace()
        }
        val width = result!!.width
        val height = result.height
        val pixels = IntArray(width * height)
        // All are 0, or black, by default
        for (y in 0 until height) {
            val offset = y * width
            for (x in 0 until width) {
                pixels[offset + x] = if (result[x, y]) codeColor else backgroundColor
            }
        }
        val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        bitmap.setPixels(pixels, 0, width, 0, 0, width, height)
        return bitmap
    }
    //----------------------------------------------------------------------------------------------生成条形码开始
    /**
     * 生成条形码
     *
     * @param contents      需要生成的内容
     * @param desiredWidth  生成条形码的宽带
     * @param desiredHeight 生成条形码的高度
     * @return
     */
    /**
     * 生成条形码
     * desiredWidth  生成条形码的宽带
     * desiredHeight 生成条形码的高度
     *
     * @param contents 需要生成的内容
     * @return 条形码的Bitmap
     */
    @JvmOverloads
    fun createBarCode(contents: String, desiredWidth: Int = 1000, desiredHeight: Int = 300): Bitmap {
        return createBarCode(contents, desiredWidth, desiredHeight, -0x1000000, -0x1)
    }

    fun createBarCode(content: String, codeWidth: Int, codeHeight: Int, iv_code: ImageView) {
        iv_code.setImageBitmap(createBarCode(content, codeWidth, codeHeight))
    }

    fun createBarCode(content: String, iv_code: ImageView) {
        iv_code.setImageBitmap(createBarCode(content))
    } //==============================================================================================生成条形码结束

    class Builder(private val content: CharSequence) {
        private var backgroundColor = -0x1
        private var codeColor = -0x1000000
        private var codeWidth = 1000
        private var codeHeight = 300
        fun backColor(backgroundColor: Int): Builder {
            this.backgroundColor = backgroundColor
            return this
        }

        fun codeColor(codeColor: Int): Builder {
            this.codeColor = codeColor
            return this
        }

        fun codeWidth(codeWidth: Int): Builder {
            this.codeWidth = codeWidth
            return this
        }

        fun codeHeight(codeHeight: Int): Builder {
            this.codeHeight = codeHeight
            return this
        }

        fun into(imageView: ImageView?): Bitmap {
            val bitmap = createBarCode(content, codeWidth, codeHeight, backgroundColor, codeColor)
            imageView?.setImageBitmap(bitmap)
            return bitmap
        }

    }
}